src/bun.js/bindings/NodeVMSourceTextModule.cpp

```diff
@@ -0,0 +1,461 @@
+#include "NodeVMSourceTextModule.h"
+
+#include "ErrorCode.h"
+#include "JSDOMExceptionHandling.h"
+
+#include "wtf/Scope.h"
+
+#include "JavaScriptCore/JIT.h"
+#include "JavaScriptCore/JSModuleRecord.h"
+#include "JavaScriptCore/JSPromise.h"
+#include "JavaScriptCore/JSSourceCode.h"
+#include "JavaScriptCore/ModuleAnalyzer.h"
+#include "JavaScriptCore/ModuleProgramCodeBlock.h"
+#include "JavaScriptCore/Parser.h"
+#include "JavaScriptCore/SourceCodeKey.h"
+#include "JavaScriptCore/Watchdog.h"
+
+#include "../vm/SigintWatcher.h"
+
+namespace Bun {
+using namespace NodeVM;
+
+NodeVMSourceTextModule* NodeVMSourceTextModule::create(VM& vm, JSGlobalObject* globalObject, ArgList args)
+{
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSValue identifierValue = args.at(0);
+    if (!identifierValue.isString()) {
+        throwArgumentTypeError(*globalObject, scope, 0, "identifier"_s, "Module"_s, "Module"_s, "string"_s);
+        return nullptr;
+    }
+
+    JSValue contextValue = args.at(1);
+    if (contextValue.isUndefined()) {
+        contextValue = globalObject;
+    } else if (!contextValue.isObject()) {
+        throwArgumentTypeError(*globalObject, scope, 1, "context"_s, "Module"_s, "Module"_s, "object"_s);
+        return nullptr;
+    }
+
+    JSValue sourceTextValue = args.at(2);
+    if (!sourceTextValue.isString()) {
+        throwArgumentTypeError(*globalObject, scope, 2, "sourceText"_s, "Module"_s, "Module"_s, "string"_s);
+        return nullptr;
+    }
+
+    JSValue lineOffsetValue = args.at(3);
+    if (!lineOffsetValue.isUInt32AsAnyInt()) {
+        throwArgumentTypeError(*globalObject, scope, 3, "lineOffset"_s, "Module"_s, "Module"_s, "number"_s);
+        return nullptr;
+    }
+
+    JSValue columnOffsetValue = args.at(4);
+    if (!columnOffsetValue.isUInt32AsAnyInt()) {
+        throwArgumentTypeError(*globalObject, scope, 4, "columnOffset"_s, "Module"_s, "Module"_s, "number"_s);
+        return nullptr;
+    }
+
+    JSValue cachedDataValue = args.at(5);
+    WTF::Vector<uint8_t> cachedData;
+    if (!cachedDataValue.isUndefined() && !extractCachedData(cachedDataValue, cachedData)) {
+        Bun::ERR::INVALID_ARG_TYPE(scope, globalObject, "options.cachedData"_s, "Buffer, TypedArray, or DataView"_s, cachedDataValue);
+        return nullptr;
+    }
+
+    uint32_t lineOffset = lineOffsetValue.toUInt32(globalObject);
+    uint32_t columnOffset = columnOffsetValue.toUInt32(globalObject);
+
+    Ref<StringSourceProvider> sourceProvider = StringSourceProvider::create(sourceTextValue.toWTFString(globalObject), SourceOrigin {}, String {}, SourceTaintedOrigin::Untainted,
```

Comment:
Does the SourceOrigin need to be set here too?

```suggestion
    auto&& sourceCodeString = sourceTextValue.toWTFString(globalObject);
    RETURN_IF_EXCEPTION(scope, nullptr);
    Ref<StringSourceProvider> sourceProvider = StringSourceProvider::create(sourceCodeString, SourceOrigin {}, String {}, SourceTaintedOrigin::Untainted,
```